internal/vendors/downloader.go

```diff
@@ -367,3 +370,64 @@ func (s *S3Downloader) Download(ctx context.Context, downloadDir string, firmwar
 
 	return path.Join(downloadDir, firmware.Filename), nil
 }
+
+// SourceOverrideDownloader is meant to download firmware from an alternate source
+// than the firmware's UpstreamURL.
+type SourceOverrideDownloader struct {
+	logger  *logrus.Logger
+	client  serverservice.Doer
+	baseURL string
+}
+
+// NewSourceOverrideDownloader creates a SourceOverrideDownloader.
+func NewSourceOverrideDownloader(logger *logrus.Logger, client serverservice.Doer, sourceURL string) Downloader {
+	if !strings.HasSuffix(sourceURL, "/") {
+		sourceURL += "/"
+	}
+
+	return &SourceOverrideDownloader{
+		logger,
+		client,
+		sourceURL,
+	}
+}
+
+// Download will download the given firmware into the given downloadDir,
+// and return the full path to the downloaded file.
+// The file will be downloaded from the sourceURL provided to the SourceOverrideDownloader
+// instead of the firmware's UpstreamURL.
+func (d *SourceOverrideDownloader) Download(ctx context.Context, downloadDir string, firmware *serverservice.ComponentFirmwareVersion) (string, error) {
```

Comment:
Can we have a test for this method added

Reply:
Sure thing! Done